plugins/aws/provisioner.go

```diff
@@ -2,42 +2,155 @@ package aws
+	if len(out.Diagnostics.Errors) > 0 {
+		return
 	}
+
+	provisioner.Provision(ctx, in, out)
 }
 
 func (p awsProvisioner) Deprovision(ctx context.Context, in sdk.DeprovisionInput, out *sdk.DeprovisionOutput) {
-	// Nothing to do here: environment variables get wiped automatically when the process exits.
+	p.selectedProvisioner.Deprovision(ctx, in, out)
 }
 
 func (p awsProvisioner) Description() string {
-	return p.envVarProvisioner.Description()
+	if p.selectedProvisioner != nil {
+		return p.selectedProvisioner.Description()
+	}
+	return "Provisions AWS with temporary credentials."
+}
+
+func (p *awsProvisioner) selectCorrespondingProvisioner(itemFields map[sdk.FieldName]string, out *sdk.ProvisionOutput) sdk.Provisioner {
+	awsConfig := getAWSConfiguration(out)
+	if len(out.Diagnostics.Errors) > 0 {
+		return nil
+	}
+
+	err := errorIfUnsupportedConfigurationIsPresent(awsConfig)
+	if err != nil {
+		out.AddError(err)
+		return nil
+	}
+
+	selectedProvisioner := selectProvisionerBasedOnPresentValues(itemFields, awsConfig, out)
+
+	p.selectedProvisioner = selectedProvisioner
+
+	return selectedProvisioner
+}
+
+func getAWSConfiguration(out *sdk.ProvisionOutput) *confighelpers.Config {
+	for i, arg := range out.CommandLine {
+		if arg == "--profile" {
+			if i+1 == len(out.CommandLine) {
+				out.AddError(fmt.Errorf("--profile flag was specified without a value"))
+				return nil
+			}
+			config := getAWSAuthConfigurationForProfile(out.CommandLine[i+1], out)
+			// Remove --profile flag so aws cli doesn't attempt to assume role by itself
+			out.CommandLine = removeProfileFlagFromArgs(i, out.CommandLine)
+			return config
+		}
+	}
+
+	if val := os.Getenv("AWS_PROFILE"); val != "" {
+		return getAWSAuthConfigurationForProfile(val, out)
+	}
+
+	return getAWSAuthConfigurationForProfile("default", out)
+}
+
+func getAWSAuthConfigurationForProfile(profile string, out *sdk.ProvisionOutput) *confighelpers.Config {
+	// Read config file from the location set in AWS_CONFIG_FILE env var or from  ~/.aws/config
+	configFile, err := confighelpers.LoadConfigFromEnv()
+	if err != nil {
+		out.AddError(err)
+		return nil
+	}
+
+	configLoader := confighelpers.ConfigLoader{
+		File:          configFile,
+		ActiveProfile: profile,
+	}
+
+	configuration, err := configLoader.LoadFromProfile(profile)
+	if err != nil {
+		out.AddError(err)
+		return nil
+	}
+
+	return configuration
+}
+
+func selectProvisionerBasedOnPresentValues(itemFields map[sdk.FieldName]string, awsConfig *confighelpers.Config, out *sdk.ProvisionOutput) sdk.Provisioner {
+	totp, hasTotp := itemFields[fieldname.OneTimePassword]
+	mfaSerial, hasMFASerial := itemFields[fieldname.MFASerial]
+
+	// Give priority to the mfa serial specified in 1Password
+	if hasMFASerial && awsConfig.HasMfaSerial() && awsConfig.MfaSerial != mfaSerial {
+		out.AddError(fmt.Errorf("your local AWS configuration has a different MFA serial than the one specified in 1Password"))
```

Comment:
It feels inconsistent that [we use the region from the 1Password item](https://github.com/1Password/shell-plugins/pull/180/files#diff-4899eeb666d138152378c9be8b70bcee7287a247ab1e504a3385044550017794R56) when both the config file and the item in 1Password have a region specified, but when both the config and the 1Password item have an MFA serial defined, there is no conflict resolution and instead an error is thrown.

To make the product work intuitively, we need the behavior to be consistent. I have a preference for raising an error in both situations, as I don't see a reason why a user would want to have different definitions in both places, so it's probably a user error. Having the product throw an error in this case is helpful, because it allows the user to fix the conflict and choose which of the two they actually meant to use.

Reply:
Sounds good. I am happy about having this as a global rule for resolving this type of conflicts!